### ISSUE TITLE
[java] EqualsMissingOnComparable: 'Comparable' implemented but 'equals()' not overridden

### ISSUE THREAD

**Issue by Pankraz76** [pmd/pmd#5837]
**EqualsMissingOnComparable:**

**Proposed Category:** One of [Best Practices | Error Prone]

might be be important. IDEA is reporting as Probable bugs.

**Description:**

collection will violate the contract of `java.util.Set`, which is defined in terms of equals() if equals() is not overridden, the equals() implementation is not consistent with the compareTo() implementation. If an object of such a class is added to a collection such as `java.util.SortedSet`

- https://www.jetbrains.com/help/inspectopedia/ComparableImplementedButEqualsNotOverridden.html

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
class Foo implements Comparable<Length> {
  @Override
  public int compareTo(@NotNull Length o) {
    return 0;
  }
}
```

```java
class Foo implements Comparable<Length> {
  @Override
  public int compareTo(@NotNull Length o) {
    return 0;
  }

  @Override
  public boolean equals(Object o) {
    return o instanceof Length && compareTo((Length) o) == 0;
  }
}
```

**Possible Properties:**

*   Should this rule be customizable via properties?

**Comment by Pankraz76** [comment]
FYI: 

hi @bclozel,
might not be important, just wanted to let you know, so you can check for yourself.





<img width="1879" alt="Image" src="https://github.com/user-attachments/assets/c1d23b76-bbba-4f56-a0f3-57ec06615b2f" />

**Comment by adangel** [comment]
More detailed explanation why this is a problem is in Effective Java, Item 12: Consider implementing `Comparable`.

This new rule sounds similar to [OverrideBothEqualsAndHashcode](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#overridebothequalsandhashcode) (although the explantation is missing there - invalid behavior of hash-based collections... See Item 9: Always override hashCode when you override equals | Effective Java)
-> I'd suggest category error prone

**Comment by Pankraz76** [comment]
right, when doing equals, hashcode is part of the contract too.

As it's about equivalence, this might be a derivative of the mentioned rule and therefore justifiably allowed to exist there, without need to dedicate an separate item, since it has very similar semantics.

What do you think?

**Comment by Pankraz76** [comment]
> Correct:

not sure about this, as missing `hashCode()`.

This way we fail 2 timesm, need one iteration more as user will fix this and then ran into the [OverrideBothEqualsAndHashcode](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#overridebothequalsandhashcode) issue, which then has to be fixed again.

We could argue therefore, to avoid this overhead, placing it straight away into the corresponding class.

**Comment by Pankraz76** [comment]
actually the `Comparable` interface itself, should include these 2 methods, enforcing equality which seems to be required when comparing.

**Comment by adangel** [comment]
> This way we fail 2 timesm, need one iteration more as user will fix this and then ran into the [OverrideBothEqualsAndHashcode](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#overridebothequalsandhashcode) issue, which then has to be fixed again.

Nobody said, that programming is easy. PMD is just a tool that helps to not forget something. I think, the behavior is acceptable.

**Comment by Pankraz76** [comment]
> This way we fail 2 timesm, need one iteration more as user will fix this and then ran into the [OverrideBothEqualsAndHashcode](https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#overridebothequalsandhashcode) issue, which then has to be fixed again.

it seems my assumption is wrong, they should not collide.

**Comment by adangel** [comment]
I'm suggesting to name this new rule "OverrideEqualsOnComparable", so that it is closer to the related existing rule "OverrideBothEqualsAndHashcode".

**Comment by Pankraz76** [comment]
agree, thanks for suggestion.
